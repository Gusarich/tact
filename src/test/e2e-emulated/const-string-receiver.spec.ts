import { toNano } from "@ton/core";
import { ContractSystem } from "@tact-lang/emulator";
import { __DANGER_resetNodeId } from "../../grammar/ast";
import { ConstStringReceiverTester } from "../e2e-emulated/contracts/output/const-string-receiver_ConstStringReceiverTester";

describe("constant string receiver", () => {
    beforeEach(() => {
        __DANGER_resetNodeId();
    });
    it("should implement const string receiver correctly", async () => {
        // Init
        const system = await ContractSystem.create();
        const treasure = system.treasure("treasure");
        const contract = system.open(
            await ConstStringReceiverTester.fromInit(),
        );

        expect(contract.abi).toMatchSnapshot();

        // Deploy
        await contract.send(treasure, { value: toNano("10") }, "string 1");
        await system.run();

        const tracker = system.track(contract);

        await contract.send(treasure, { value: toNano("10") }, "string 1");
        await contract.send(treasure, { value: toNano("10") }, "string 2");
        await contract.send(treasure, { value: toNano("10") }, "string 3");
        await system.run();

        expect(tracker.collect()).toMatchSnapshot();
    });
});
