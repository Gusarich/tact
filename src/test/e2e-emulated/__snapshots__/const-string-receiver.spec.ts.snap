// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`constant string receiver should implement const string receiver correctly 1`] = `
{
  "errors": {
    "10": {
      "message": "Dictionary error",
    },
    "128": {
      "message": "Null reference exception",
    },
    "129": {
      "message": "Invalid serialization prefix",
    },
    "13": {
      "message": "Out of gas error",
    },
    "130": {
      "message": "Invalid incoming message",
    },
    "131": {
      "message": "Constraints error",
    },
    "132": {
      "message": "Access denied",
    },
    "133": {
      "message": "Contract stopped",
    },
    "134": {
      "message": "Invalid argument",
    },
    "135": {
      "message": "Code of a contract was not found",
    },
    "136": {
      "message": "Invalid address",
    },
    "137": {
      "message": "Masterchain support is not enabled for this contract",
    },
    "2": {
      "message": "Stack underflow",
    },
    "3": {
      "message": "Stack overflow",
    },
    "32": {
      "message": "Method ID not found",
    },
    "34": {
      "message": "Action is invalid or not supported",
    },
    "37": {
      "message": "Not enough TON",
    },
    "38": {
      "message": "Not enough extra-currencies",
    },
    "4": {
      "message": "Integer overflow",
    },
    "5": {
      "message": "Integer out of expected range",
    },
    "6": {
      "message": "Invalid opcode",
    },
    "7": {
      "message": "Type check error",
    },
    "8": {
      "message": "Cell overflow",
    },
    "9": {
      "message": "Cell underflow",
    },
  },
  "getters": [],
  "receivers": [
    {
      "message": {
        "kind": "text",
        "text": "string 1",
      },
      "receiver": "internal",
    },
    {
      "message": {
        "kind": "text",
        "text": "string 2",
      },
      "receiver": "internal",
    },
    {
      "message": {
        "kind": "text",
        "text": "string 3",
      },
      "receiver": "internal",
    },
  ],
  "types": [
    {
      "fields": [
        {
          "name": "code",
          "type": {
            "kind": "simple",
            "optional": false,
            "type": "cell",
          },
        },
        {
          "name": "data",
          "type": {
            "kind": "simple",
            "optional": false,
            "type": "cell",
          },
        },
      ],
      "header": null,
      "name": "StateInit",
    },
    {
      "fields": [
        {
          "name": "bounced",
          "type": {
            "kind": "simple",
            "optional": false,
            "type": "bool",
          },
        },
        {
          "name": "sender",
          "type": {
            "kind": "simple",
            "optional": false,
            "type": "address",
          },
        },
        {
          "name": "value",
          "type": {
            "format": 257,
            "kind": "simple",
            "optional": false,
            "type": "int",
          },
        },
        {
          "name": "raw",
          "type": {
            "kind": "simple",
            "optional": false,
            "type": "slice",
          },
        },
      ],
      "header": null,
      "name": "Context",
    },
    {
      "fields": [
        {
          "name": "bounce",
          "type": {
            "kind": "simple",
            "optional": false,
            "type": "bool",
          },
        },
        {
          "name": "to",
          "type": {
            "kind": "simple",
            "optional": false,
            "type": "address",
          },
        },
        {
          "name": "value",
          "type": {
            "format": 257,
            "kind": "simple",
            "optional": false,
            "type": "int",
          },
        },
        {
          "name": "mode",
          "type": {
            "format": 257,
            "kind": "simple",
            "optional": false,
            "type": "int",
          },
        },
        {
          "name": "body",
          "type": {
            "kind": "simple",
            "optional": true,
            "type": "cell",
          },
        },
        {
          "name": "code",
          "type": {
            "kind": "simple",
            "optional": true,
            "type": "cell",
          },
        },
        {
          "name": "data",
          "type": {
            "kind": "simple",
            "optional": true,
            "type": "cell",
          },
        },
      ],
      "header": null,
      "name": "SendParameters",
    },
  ],
}
`;

exports[`constant string receiver should implement const string receiver correctly 2`] = `
[
  {
    "$seq": 1,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "text": "string 1",
            "type": "text",
          },
          "bounce": true,
          "from": "@treasure(treasure)",
          "to": "kQDg3ce4axASb2iP16WY4k7nP9uV--zZxPfmGYbm8ZF_YPj4",
          "type": "internal",
          "value": "10",
        },
      },
      {
        "$type": "processed",
        "gasUsed": 6689n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "text": "string 1",
              "type": "text",
            },
            "bounce": true,
            "from": "kQDg3ce4axASb2iP16WY4k7nP9uV--zZxPfmGYbm8ZF_YPj4",
            "to": "@treasure(treasure)",
            "type": "internal",
            "value": "9.992115",
          },
        ],
      },
    ],
  },
  {
    "$seq": 2,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "text": "string 2",
            "type": "text",
          },
          "bounce": true,
          "from": "@treasure(treasure)",
          "to": "kQDg3ce4axASb2iP16WY4k7nP9uV--zZxPfmGYbm8ZF_YPj4",
          "type": "internal",
          "value": "10",
        },
      },
      {
        "$type": "processed",
        "gasUsed": 9300n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "text": "string 2",
              "type": "text",
            },
            "bounce": true,
            "from": "kQDg3ce4axASb2iP16WY4k7nP9uV--zZxPfmGYbm8ZF_YPj4",
            "to": "@treasure(treasure)",
            "type": "internal",
            "value": "9.989504",
          },
        ],
      },
    ],
  },
  {
    "$seq": 3,
    "events": [
      {
        "$type": "received",
        "message": {
          "body": {
            "text": "string 3",
            "type": "text",
          },
          "bounce": true,
          "from": "@treasure(treasure)",
          "to": "kQDg3ce4axASb2iP16WY4k7nP9uV--zZxPfmGYbm8ZF_YPj4",
          "type": "internal",
          "value": "10",
        },
      },
      {
        "$type": "processed",
        "gasUsed": 9377n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "text": "string 3",
              "type": "text",
            },
            "bounce": true,
            "from": "kQDg3ce4axASb2iP16WY4k7nP9uV--zZxPfmGYbm8ZF_YPj4",
            "to": "@treasure(treasure)",
            "type": "internal",
            "value": "9.989427",
          },
        ],
      },
    ],
  },
]
`;
